### __FEATURE SELECTION FOR MACHINE LEARNING__


_Attention : in order to re-run notebook you need to put DLBCL and Prostate files together with notebook in the same directory_

Research conducted with following algorithms: 

1.  Scoring fucntion: ANOVA
2.  SVM classifier with recursive feature elimination and crossvalidation 
3.  Decision Tree and RandomForest classifiers 

Parameter modified in the study was number of features. Parameter was included in the algoritms as follows:
1. Anova : returns number of features with the highest numbers of points
2. RFE SVM : number of features means minimum number of features returned by aglorithm ("depth of analysis")
3. Decision trees : maximum number of featuresmaksymalna to be included in tree split

Study conducted for the following parameters:   
number of features = 0,1%,0.5%,1%,5%,10%,15%,20%,30%,50% of all features in the datasets.

Results of the algorithms, which is quality of the selected feature set, was measured with mean square error of classification using:
NBC,5NN,J48. Additionaly computation time was measured.
 



In [2]:
import selekcja_cech as sc
import pandas as pd

In [3]:
procent_atr=[0.1,0.5,1,5,10,15,20,30,50]
dlbcl_training = 'dlbcl_training.csv'
dlbcl_test ='dlbcl_testing.csv'

pro_training = 'prostate_training.csv'
pro_test ='prostate_testing.csv'


X_train,y_train,X_test,y_test = sc.load_data(dlbcl_training,dlbcl_test)
dlbcl=sc.select_feature(X_train, y_train, X_test, y_test, procent_atr)

X_train,y_train,X_test,y_test = sc.load_data(pro_training,pro_test)
prostate=sc.select_feature(X_train, y_train, X_test, y_test, procent_atr)

### 1. __RESULTS__

#### 1.1 Best selected features set:<br>
    Due to the identical results of selection quality('Error') second criterion is the samllest size of the feature set('N.features')

DLBCL : 

In [5]:
df=dlbcl.sort_values(by=['Błąd','L.cech'])
df.columns=['Algorithm','Error','N.features','Calc time','% of features']
df.head(1)

Algorithm  Error  N.features  Calc time  % of features
24      Tree  0.104           3      0.063           20.0

PROSTATE :

In [6]:
df=prostate.sort_values(by=['Błąd','L.cech'])
df.columns=['Algorithm','Error','N.features','Calc time','% of features']
df.head(1)

Algorithm  Error  N.features  Calc time  % of features
9       RFE  0.133           2      4.541            0.1

#### 1.2 10 results with the smallest errors

In [7]:
d=dlbcl.sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
d.columns=['Algorithm','Error','N.features','Calc time','% of features']
p=prostate.sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
p.columns=['Algorithm','Error','N.features','Calc time','% of features']
pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)

DLBCL                                            PROSTATE         \
  Algorithm  Error N.features Calc time % of features Algorithm  Error   
0      Tree  0.104          3     0.063          20.0       RFE  0.133   
1      Tree  0.104          4     0.059          15.0     Anova  0.133   
2     Anova  0.104        132     0.067           5.0     Anova  0.133   
3     Anova  0.104        397     0.073          15.0      Tree  0.150   
4      Tree  0.125          3     0.069          30.0     Anova  0.150   
5    Random  0.125        158     1.011          30.0       RFE  0.150   
6    Random  0.125        183     0.795          20.0     Anova  0.150   
7    Random  0.125        231     0.523           5.0       RFE  0.150   
8    Random  0.146        161     1.394          50.0     Anova  0.150   
9    Random  0.146        187     0.717          15.0    Random  0.150   

                                      
  N.features Calc time % of features  
0          2     4.541           0.1  
1        214     0.061          10.0  
2        320     0.101          15.0  
3          4     0.073          50.0  
4         11     0.056           0.5  
5         11     4.267           0.5  
6         21     0.058           1.0  
7         21     4.356           1.0  
8        107     0.057           5.0  
9        157     1.603          50.0

#### 1.3 Best results for each methods:

In [20]:
d=dlbcl.sort_values(['Błąd','L.cech']).groupby('Algorytm').first().sort_values(['Błąd','L.cech']).reset_index()
d.columns=['Algorithm','Error','N.features','Calc time','% of features']
d.loc[d['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

p=prostate.sort_values(['Błąd','L.cech']).groupby('Algorytm').first().sort_values(['Błąd','L.cech']).reset_index()
p.columns=['Algorithm','Error','N.features','Calc time','% of features']
p.loc[p['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)


DLBCL                                                 PROSTATE  \
      Algorithm   Error N.features Calc time % of features     Algorithm   
0          Tree  0.1040          3     0.063          20.0           RFE   
1         Anova  0.1040        132     0.067           5.0         Anova   
2        Random  0.1250        158     1.011          30.0          Tree   
3           RFE  0.1670       1324     3.053          50.0        Random   
4  No selection  0.1875          0     0.000           0.0  No selection   

                                             
   Error N.features Calc time % of features  
0  0.133          2     4.541           0.1  
1  0.133        214     0.061          10.0  
2  0.150          4     0.073          50.0  
3  0.150        157     1.603          50.0  
4  0.200          0     0.000           0.0


Each algorithms for each datasets gave better results than classification with no feature selection (no selection). Results indicate a relationship between algorithm and dataset.


### 2. __Errors range__ :

In [88]:

print('DLBCL: <',
      round(dlbcl.sort_values(by=['Błąd','L.cech']).head(1).Błąd.values[0]*100,3),
      '%,',
      round(dlbcl.sort_values(by=['Błąd','L.cech']).tail(1).Błąd.values[0]*100,3),'% >')
print('PROSTATE: <',
      round(prostate.sort_values(by=['Błąd','L.cech']).head(1).Błąd.values[0]*100,3),
      '%,',
      round(prostate.sort_values(by=['Błąd','L.cech']).tail(1).Błąd.values[0]*100,3),'% >')

DLBCL: < 10.4 %, 41.7 % >
PROSTATE: < 13.3 %, 33.3 % >


 
Errors range is huge. Results don't show clear relation between error and number of features and datasets.

### 3. __ALGORITHMS__

#### 3.1 ANOVA

In [21]:
d=dlbcl[dlbcl['Algorytm'].isin(['Anova','Bez_selekcji'])].sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
d.columns=['Algorithm','Error','N.features','Calc time','% of features']
d.loc[d['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

p=prostate[prostate['Algorytm'].isin(['Anova','Bez_selekcji'])].sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
p.columns=['Algorithm','Error','N.features','Calc time','% of features']
p.loc[p['Algorithm']=='Bez_selekcji','Algorithm']='No selection'


pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)

DLBCL                                                 PROSTATE  \
      Algorithm   Error N.features Calc time % of features     Algorithm   
0         Anova  0.1040        132     0.067           5.0         Anova   
1         Anova  0.1040        397     0.073          15.0         Anova   
2         Anova  0.1460        265     0.079          10.0         Anova   
3         Anova  0.1460        529     0.095          20.0         Anova   
4         Anova  0.1460        794     0.096          30.0         Anova   
5  No selection  0.1875          0     0.000           0.0         Anova   
6         Anova  0.1880         26     0.087           1.0         Anova   
7         Anova  0.2080       1324     0.118          50.0         Anova   
8         Anova  0.2290         13     0.073           0.5  No selection   
9         Anova  0.2710          3     0.086           0.1         Anova   

                                             
   Error N.features Calc time % of features  
0  0.133        214     0.061          10.0  
1  0.133        320     0.101          15.0  
2  0.150         11     0.056           0.5  
3  0.150         21     0.058           1.0  
4  0.150        107     0.057           5.0  
5  0.150        427     0.088          20.0  
6  0.150        640     0.083          30.0  
7  0.150       1068     0.089          50.0  
8  0.200          0     0.000           0.0  
9  0.200          2     0.057           0.1

Anova gives good or very good results (low Errors) no matter what datasets used for calculation.Drawback is relativly big dimension of selected features set.Algorithm is fast.  Best three results was generated with less than 15% of features.

#### 3.2 RFE

In [22]:
d=dlbcl[dlbcl['Algorytm'].isin(['RFE','Bez_selekcji'])].sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
d.columns=['Algorithm','Error','N.features','Calc time','% of features']
d.loc[d['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

p=prostate[prostate['Algorytm'].isin(['RFE','Bez_selekcji'])].sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
p.columns=['Algorithm','Error','N.features','Calc time','% of features']
p.loc[p['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)

DLBCL                                                 PROSTATE  \
      Algorithm   Error N.features Calc time % of features     Algorithm   
0           RFE  0.1670       1324     3.053          50.0           RFE   
1  No selection  0.1875          0     0.000           0.0           RFE   
2           RFE  0.1880        265     4.578          10.0           RFE   
3           RFE  0.1880        529     3.771          20.0           RFE   
4           RFE  0.2080        397     4.072          15.0           RFE   
5           RFE  0.2080        794     3.683          30.0           RFE   
6           RFE  0.2290        132     4.833           5.0           RFE   
7           RFE  0.2500         67     5.262           0.1           RFE   
8           RFE  0.2500         67     4.358           0.5           RFE   
9           RFE  0.2500         67     4.340           1.0  No selection   

                                             
   Error N.features Calc time % of features  
0  0.133          2     4.541           0.1  
1  0.150         11     4.267           0.5  
2  0.150         21     4.356           1.0  
3  0.167        795     3.967           5.0  
4  0.167        795     3.911          10.0  
5  0.167        795     3.818          15.0  
6  0.167        795     3.840          20.0  
7  0.167        795     3.598          30.0  
8  0.183       1145     2.953          50.0  
9  0.200          0     0.000           0.0

RFE: results strongly depend on dataset. For DBCL set, except one case (50% od features selected), all results were worse than 'no selection' classification. For PROSTATE set all results were better then 'no selection' classification. Best RFE results for PROSTATE file beat other algorithms with low error and low number of selected features ('Błąd'=0,133, RFE : 2 features , ANOVA:214 features).The visible relationship between the number of features and the obtained result: inversely proportional in the case of a matched data set (PROSTATE, the deeper the algorithm goes, the better the result is) and almost directly proportional when the data set is not matched (DBCL: the deeper it goes the results is worse). Computationally expensive algorithm. Selection time two orders of magnitude greater than Anova and Decision Tree.

#### 3.3 DecisionTREE/RandomForest

In [23]:
d=dlbcl[dlbcl['Algorytm'].isin(['Tree','Bez_selekcji'])].sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
d.columns=['Algorithm','Error','N.features','Calc time','% of features']
d.loc[d['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

p=prostate[prostate['Algorytm'].isin(['Tree','Bez_selekcji'])].sort_values(by=['Błąd','L.cech']).head(10).reset_index(drop=True)
p.columns=['Algorithm','Error','N.features','Calc time','% of features']
p.loc[p['Algorithm']=='Bez_selekcji','Algorithm']='No selection'

pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)

DLBCL                                                 PROSTATE  \
      Algorithm   Error N.features Calc time % of features     Algorithm   
0          Tree  0.1040          3     0.063          20.0          Tree   
1          Tree  0.1040          4     0.059          15.0          Tree   
2          Tree  0.1250          3     0.069          30.0          Tree   
3  No selection  0.1875          0     0.000           0.0  No selection   
4          Tree  0.2080          4     0.059          10.0          Tree   
5          Tree  0.2500          3     0.059           5.0          Tree   
6          Tree  0.2500          7     0.060           0.5          Tree   
7          Tree  0.2710          3     0.084          50.0          Tree   
8          Tree  0.3750          5     0.058           1.0          Tree   
9          Tree  0.4170          8     0.066           0.1          Tree   

                                             
   Error N.features Calc time % of features  
0  0.150          4     0.073          50.0  
1  0.167          5     0.069          30.0  
2  0.183         12     0.048           1.0  
3  0.200          0     0.000           0.0  
4  0.200          7     0.056           0.5  
5  0.217          8     0.057          15.0  
6  0.217          9     0.053           5.0  
7  0.233          6     0.049          10.0  
8  0.233          6     0.058          20.0  
9  0.333         14     0.065           0.1

DecisionTree / RandomForest: As with RFE, the results obtained depend on the dataset. For DBCL: 8 of 10 best results are DecisionTree or RandomForest, for PROSTATE only 1 of 10 best results is  generated by DecisionTree. DecisionTree selects small feature subsets, usually producing the best results with larger values of % of features parameter, while RandomForest produces more numerous feature subsets.DT produces the highest error range of any algorithm (DBCL: 10.4% - 41.7%, PROSTATE: 15% -33 %) DT is a fast algorithm.

### __4.TIME__

#### 4.1 10 results with best computation time 

In [24]:
d=dlbcl[~(dlbcl['Algorytm']=='Bez_selekcji')].sort_values(by=['Czas','Błąd']).head(10).reset_index(drop=True)
d.columns=['Algorithm','Error','N.features','Calc time','% of features']
p=prostate[~(prostate['Algorytm']=='Bez_selekcji')].sort_values(by=['Czas','Błąd']).head(10).reset_index(drop=True)
p.columns=['Algorithm','Error','N.features','Calc time','% of features']
pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)

DLBCL                                            PROSTATE         \
  Algorithm  Error N.features Calc time % of features Algorithm  Error   
0      Tree  0.375          5     0.058           1.0      Tree  0.183   
1      Tree  0.104          4     0.059          15.0      Tree  0.233   
2      Tree  0.208          4     0.059          10.0      Tree  0.217   
3      Tree  0.250          3     0.059           5.0     Anova  0.150   
4      Tree  0.250          7     0.060           0.5      Tree  0.200   
5      Tree  0.104          3     0.063          20.0     Anova  0.150   
6      Tree  0.417          8     0.066           0.1     Anova  0.200   
7     Anova  0.104        132     0.067           5.0      Tree  0.217   
8      Tree  0.125          3     0.069          30.0     Anova  0.150   
9     Anova  0.104        397     0.073          15.0      Tree  0.233   

                                      
  N.features Calc time % of features  
0         12     0.048           1.0  
1          6     0.049          10.0  
2          9     0.053           5.0  
3         11     0.056           0.5  
4          7     0.056           0.5  
5        107     0.057           5.0  
6          2     0.057           0.1  
7          8     0.057          15.0  
8         21     0.058           1.0  
9          6     0.058          20.0

#### 4.2 Selection mean time

In [34]:
d=dlbcl.groupby(by='Algorytm').mean().Czas
d.index.name='Algorithm'
p=prostate.groupby(by='Algorytm').mean().Czas
p.index.name='Algorithm'
pd.concat([d,p],keys=['DLBCL','PROSTATE'],axis=1)



DLBCL  PROSTATE
Algorithm                       
Anova         0.086000  0.072222
Bez_selekcji  0.000000  0.000000
RFE           4.216667  3.916778
Random        0.690556  0.763222
Tree          0.064111  0.058667

The selection time is closely related to the complexity of the algorithm and in principle does not depend on either the data set or the parameter % of features.